File: CSharp/One/Transforms/CollectInheritanceInfo.cs
using One.Ast;
using One;

namespace One.Transforms
{
    public class CollectInheritanceInfo : ITransformer {
        public string name { get; set; }
        
        public CollectInheritanceInfo()
        {
            // C# fix
            this.name = "CollectInheritanceInfo";
        }
        
        public void visitClass(Class cls)
        {
            var allBaseIIntfs = cls.getAllBaseInterfaces();
            var intfs = allBaseIIntfs.map(x => x is Interface int_ ? int_ : null).filter(x => x != null);
            var clses = allBaseIIntfs.map(x => x is Class class_ ? class_ : null).filter(x => x != null && x != cls);
            
            foreach (var field in cls.fields)
                field.interfaceDeclarations = intfs.map(x => x.fields.find(f => f.name == field.name)).filter(x => x != null);
            
            foreach (var method in cls.methods) {
                method.interfaceDeclarations = intfs.map(x => x.methods.find(m => m.name == method.name)).filter(x => x != null);
                method.overrides = clses.map(x => x.methods.find(m => m.name == method.name)).find(x => x != null);
                if (method.overrides != null)
                    method.overrides.overriddenBy.push(method);
            }
        }
        
        public void visitFiles(SourceFile[] files)
        {
            foreach (var file in files)
                foreach (var cls in file.classes)
                    this.visitClass(cls);
        }
    }
}

File: PHP/One/Transforms/CollectInheritanceInfo.php
<?php

namespace One\Transforms\CollectInheritanceInfo;

use One\Ast\Types\Package;
use One\Ast\Types\Class_;
use One\Ast\Types\Interface_;
use One\Ast\Types\SourceFile;
use One\ITransformer\ITransformer;

class CollectInheritanceInfo implements ITransformer {
    public $name;
    
    function __construct() {
        // C# fix
        $this->name = "CollectInheritanceInfo";
    }
    
    function visitClass($cls) {
        $allBaseIIntfs = $cls->getAllBaseInterfaces();
        $intfs = array_values(array_filter(array_map(function ($x) { return $x instanceof Interface_ ? $x : null; }, $allBaseIIntfs), function ($x) { return $x !== null; }));
        $clses = array_values(array_filter(array_map(function ($x) { return $x instanceof Class_ ? $x : null; }, $allBaseIIntfs), function ($x) use ($cls) { return $x !== null && $x !== $cls; }));
        
        foreach ($cls->fields as $field)
            $field->interfaceDeclarations = array_values(array_filter(array_map(function ($x) use ($field) { return \OneLang\ArrayHelper::find($x->fields, function ($f) use ($field) { return $f->name === $field->name; }); }, $intfs), function ($x) { return $x !== null; }));
        
        foreach ($cls->methods as $method) {
            $method->interfaceDeclarations = array_values(array_filter(array_map(function ($x) use ($method) { return \OneLang\ArrayHelper::find($x->methods, function ($m) use ($method) { return $m->name === $method->name; }); }, $intfs), function ($x) { return $x !== null; }));
            $method->overrides = \OneLang\ArrayHelper::find(array_map(function ($x) use ($method) { return \OneLang\ArrayHelper::find($x->methods, function ($m) use ($method) { return $m->name === $method->name; }); }, $clses), function ($x) { return $x !== null; });
            if ($method->overrides !== null)
                $method->overrides->overriddenBy[] = $method;
        }
    }
    
    function visitFiles($files) {
        foreach ($files as $file)
            foreach ($file->classes as $cls)
                $this->visitClass($cls);
    }
}
